Update branching.md (#182)
neondatabase/website: content/docs/conceptual-guides/branching.md

	
- **Machine Learning**
    -	Create a branch for ML model training
    -	Name or tag a branch for a specific point in time for ML model training repeatability

## Branch characteristics

A branch has the following characteristics:
- A branch is created with ?vCPU and ?GB of RAM
- Branches are currently read-write 
- Only one branch endpoint is currently supported
- A branch can be created without an endpoint
- A branch endpoint can be deleted
- The endpoint associated with the branch is deleted when a branch is deleted
- A branch endpoint URL uses the same format as a project endpoint URL: `{project_id}-{endpoint_id}.cloud.neon.tech` 

## Creating a branch

To create a branch:

1. In the Neon Console, select the **Branches** tab.

2. Click **New Branch**.

3. Enter a name for the branch.

3. Select the Neon project or brach that you want to branch from.

4. Select the type of branch you want to create. 
    - **Head**: Branch the current state of the parent project. The branch is created with all of the parent project data.
    - **Time**: Branch a specific point in time. The branch is created with the project data as it existed at the specified date and time.

>     - **LSN**: Branch from a specified Log Sequence NUmber (LSN). The branch is created with the project data as it existed at the specified LSN.

Reviewer: Should the`u` in `NUmber` be lowercase?

Author: Fixed!

>     - **LSN**: Branch from a specified Log Sequence NUmber (LSN). The branch is created with the project data as it existed at the specified LSN.

Reviewer 2: ```suggestion
    - **LSN**: Branch from a specified Log Sequence Number (LSN). The branch is created with the project data as it existed at the specified LSN.
```

### What can branching be used for?

Branching has a variety of possible uses, some of which are describes below:

- **Development**
    - Create a branch of your production database for your Development team
    - Create a branch for each Developer
	
- **Testing** 
    -	Run tests on a current branch of production data
    -	Test potentially destructive queries before deploying them to production
    -	Test schema changes
    -	Run tests on real data &mdash; branching eliminates the need to hydrate a test database
    -	Run tests in parallel on separate branches, each with its own dedicated compute

- **Staging**
    - Create a staging database by branching your production data
    - Create a branch for every pull request in your CI/CD pipeline
			
- **Backup**
    -	Quickly and easily create backup branches
    -	Instantly restore a previous state by branching from a previously created backup branch
    -	Name backup branches according to the time they were created for convenient point-in-time restore (PITR)
	

> - **Replication**

Reviewer 2: drop?

Author: dropped.

---
title: Branching
enableTableOfContents: true
redirectFrom:
  - /docs/conceptual-guides/branches
---

<a id="branches-coming-soon/"></a>

Neon allows you to instantly branch your data in the same way that you branch your code. You can quickly and cost-effectively branch your data for development, testing, staging, and various of other purposes, enabling you to improve developer productivity and optimize continuous integration and delivery (CI/CD) pipelines. See [Branching workflows](#branching-workflows) for a discussion of different ways you can integrate branching into your development workflows.

_Neon Branching capabilities are not yet publicly available. If you would like to try this feature, reach out to [iwantbranching@neon.tech](mailto:iwantbranching@neon.tech) describing your use case and requesting that Neon enable branching for your account._

## What is a branch?

> A branch is a copy-on-write clone of your data. You can create a branch from a current or past state. For example, you can create a branch that includes all data up to the current point in time or up to a past point in time or [Log Sequence Number (LSN)](../../reference/glossary#lsn).

Reviewer 2: not sure the link `../../reference/glossary#lsn` resolves for me!
do we have that header in glossary?

## Branching workflows

You can use Neon's branching feature in variety development workflows, a few of which are discussed below.

### Development

You can create a branch of your production database that developers are free to play with and modify. You can quickly create a branch with all of the data that existed in the parent branch, eliminating the setup time required to deploy and maintain a development database. Branching is so easy and cost-effective that you can create a branch for each developer. For example, you can create branches from a primary development branch to assign tasks to be worked on in parallel.

### Testing

Branching enables testers to use the most recent production data. Testers can create branches for testing schema changes, validating new queries, or testing potentially destructive queries before deploying them into production. A branch is isolated from its parent branch but has all of the parent branch's data up to the point of branch creation, which eliminates the effort involved in hydrating a database. Tests can also run on separate branches in parallel, with each branch having dedicated compute resources.

> Another testing scenario enabled by branching is tracking down corruption or data quality issue. For example, you can create and dispose of as many point-in-time database branches as necessary to determine when the corruption or data quality issue first appeared.

Reviewer 2: ```suggestion
Another testing scenario enabled by branching is tracking down corruption or data quality issue. For example, you can create and dispose of multiple point-in-time database branches to determine when the corruption or data quality issue first appeared.
```

the amount of endpoints one can have simultaneously will be limited.
there is also a limit on the amount of branches you can have.

### Staging

> With Neon's branching capabilities, you can create a staging database by branching your production database. Using the Neon API, you can automate branch creation for every pull request in your CI/CD pipeline.

Reviewer 2: ```suggestion
With Neon's branching capabilities, you can create a staging database by branching your production database. Using the Neon API, you can automate branch creation creating a database branch based on the staging branch for every pull request in your CI/CD pipeline.
```
